### PR TITLE
Bug 2019948: Improving console.extensions links and routes

### PR DESCRIPTION
**Pull request by metalice**
Signed-off-by: Matan Schatzman <mschatzm@redhat.com>

/virtualization => /kubevirt.io~{version}~VirtualMachine
/virtualization/templates => /virtualmachinetemplates

Templates are below VritualMachines

Links are rewritten to new settings.

### CODE REVIEW

File: frontend/packages/kubevirt-plugin/console-extensions.json [hunk 1]
```diff
@@ -6,184 +6,213 @@
...
+      "href": "/virtualmachinetemplates",
+      "prefixNamespaced": true
     },
     "flags": {
       "required": ["KUBEVIRT"]
     }
   },
   {
-    "type": "console.page/route/standalone",
+    "type": "console.page/route",
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/console-extensions.json:107]: should use the list page
so the route for the list page:
[route to list page] - show list
[route to list page]/~new - show new object page (yaml)
[route to list page]/~new/form - show new object page (form a.k.a wizard)

File: frontend/packages/kubevirt-plugin/console-extensions.json [hunk 2]
```diff
@@ -6,184 +6,213 @@
...
-      "required": ["KUBEVIRT"]
+      "required": ["KUBEVIRT", "KUBEVIRT_HAS_v1alpha3_API"]
     }
   },
   {
-    "type": "console.pvc/alert",
+    "type": "console.page/route",
     "properties": {
-      "alert": { "$codeRef": "pvcAlert.PVCAlertExtension" }
+      "path": ["/k8s/ns/:ns/virtualmachinetemplates/~new-wizard"],
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/console-extensions.json:135]: ~new         : yaml
~new/form : form a.k.a wizard

File: frontend/packages/kubevirt-plugin/console-extensions.json [hunk 3]
```diff
@@ -258,119 +286,139 @@
...
     "properties": {
-      "model": {
-        "group": "kubevirt.io",
-        "kind": "VirtualMachineInstance"
-      },
-      "component": {
-        "$codeRef": "VirtualMachinesInstanceDetailsPage.VirtualMachinesInstanceDetailsPage"
-      }
+      "type": "VmTemplate",
+      "title": "%kubevirt-plugin~Virtual Machines%",
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/console-extensions.json:317]: are you sure it's Virtual Machines and not VirtualMachines , what other titles we have in the catalog ?

File: frontend/packages/kubevirt-plugin/console-extensions.json [hunk 4]
```diff
@@ -258,119 +286,139 @@
...
+      "catalogDescription": "%kubevirt-plugin~Quickly create a virtual machine from a template.%",
+      "typeDescription": "%kubevirt-plugin~**Virtual Machines** are templates for quickly creating a virtual machine.%",
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/console-extensions.json:319]: Virtual Machines are templates for quickly creating a virtual machine

WTF :-) ? VMs are templates ?

can you look how we ended up with nice lyrics ? can we do something about it ?

File: frontend/packages/kubevirt-plugin/src/constants/url-params.ts [hunk 5]
```diff
@@ -1,6 +1,13 @@
+import { hasV1APIVersion } from '../flags/detect-kubevirt-virtual-machines';
+
 export enum VMWizardURLParams {
   NAMESPACE = 'namespace',
   MODE = 'mode',
   INITIAL_DATA = 'initialData',
   VIEW = 'view',
 }
+
+export const VIRTUALMACHINES_BASE_URL = `kubevirt.io~${
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/src/constants/url-params.ts:10]: we already have a method for that.
@glekner please review

File: frontend/packages/kubevirt-plugin/src/utils/utils.ts [hunk 6]
```diff
@@ -42,10 +45,10 @@ export const wrapWithProgress = (setProgress: (inProgress: boolean) => void) =>
 
 export const getVMLikeModelListPath = (isCreateTemplate: boolean, namespace: string) =>
   isCreateTemplate
-    ? `/k8s/ns/${namespace}/virtualization/templates`
-    : `/k8s/ns/${namespace}/virtualization`;
+    ? `/k8s/ns/${namespace}/${VIRTUALMACHINES_TEMPLATES_BASE_URL}`
+    : `/k8s/ns/${namespace}/${VIRTUALMACHINES_BASE_URL}`;
 
 export const getVMLikeModelDetailPath = (isTemplate: boolean, namespace: string, name: string) =>
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/src/utils/utils.ts:51]: in the .json file you have plugins that redirect from a resource to the rendered page, so you know if it's template or not when you call who ever use this method, so the path can be constant in each call

File: frontend/packages/kubevirt-plugin/src/constants/url-params.ts (hunk 5 above)
Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/constants/url-params.ts:10]: `kubevirtReferenceForModel(VirtualMachineModel)`

Comment (glekner) [reply frontend/packages/kubevirt-plugin/src/constants/url-params.ts:10]: kubevirtReferenceForModel(VirtualMachineModel)

File: frontend/packages/kubevirt-plugin/src/flags/detect-kubevirt-virtual-machines.ts [hunk 7]
```diff
@@ -7,10 +7,12 @@ import {
   FLAG_KUBEVIRT_HAS_V1ALPHA3_API,
 } from './const';
 
+/* eslint-disable import/no-mutable-exports */
+export let hasV1APIVersion = false;
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/flags/detect-kubevirt-virtual-machines.ts:11]: why these changes are here?

**Review by glekner — CHANGES_REQUESTED** [review]

File: frontend/packages/kubevirt-plugin/console-extensions.json [hunk 8]
```diff
@@ -6,184 +6,213 @@
...
+      },
+      "component": { "$codeRef": "VirtualMachinesDetailsPage.VirtualMachinesDetailsPage" }
     },
     "flags": {
       "required": ["KUBEVIRT"]
     }
   },
   {
-    "type": "console.context-provider",
+    "type": "console.navigation/href",
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/console-extensions.json:93]: lets use `console.page/resource` for templates too, 
pros: we will be using the console to create the links and routes
cons: we will need to rethink the list and details pages to not shadow none vm templates

File: frontend/packages/kubevirt-plugin/console-extensions.json [hunk 9]
```diff
@@ -6,184 +6,213 @@
...
     }
   },
   {
-    "type": "console.pvc/create-prop",
+    "type": "console.page/route",
     "properties": {
-      "label": "%kubevirt-plugin~With Data upload form%",
-      "path": "~new/upload-form"
+      "exact": true,
+      "path": ["/k8s/ns/:ns/kubevirt.io~v1alpha3~VirtualMachine/~new/form"],
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/console-extensions.json:123]: We should not have explicit paths for models we support, 
an option will be to hijack toe detail page and look for name "~new"
but the dynamic plugin may have a `console.page/...` ready made for the "~new" pages, as they are widely used.
we should investigate how other resources declare the `~new` page

Comment (yaacov) [reply frontend/packages/kubevirt-plugin/console-extensions.json:123]: @glekner @vojtechszocs do we have a dynamic plugin for "new" resource page like we have for lists and details ?

File: frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/create-vm-wizard-footer.tsx [hunk 10]
```diff
@@ -10,6 +9,7 @@ import {
 import { css } from '@patternfly/react-styles';
 import styles from '@patternfly/react-styles/css/components/Wizard/wizard';
 import * as _ from 'lodash';
+/* eslint-disable no-underscore-dangle */
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/src/components/create-vm-wizard/create-vm-wizard-footer.tsx:12]: why ?

File: frontend/packages/kubevirt-plugin/src/components/modals/menu-actions-modals/utils.ts [hunk 11]
```diff
@@ -1,11 +1,12 @@
 import { history } from '@console/internal/components/utils';
 import { VMGenericLikeEntityKind } from 'packages/kubevirt-plugin/src/types/vmLike';
+import { VIRTUALMACHINES_BASE_URL } from '../../../constants/url-params';
 import { getName, getNamespace } from '../../../selectors';
 
 export const redirectToList = (vmi: VMGenericLikeEntityKind, tab?: 'templates' | '' | null) => {
```
Comment (yaacov) [inline frontend/packages/kubevirt-plugin/src/components/modals/menu-actions-modals/utils.ts:6]: it looks strange that we need to do this after conforming with console urls and page order.
do we still need breadcrumbs (navigation is flat), do other resource pages use them ?
maybe the console have a "standard" way of creating breadcrumbs we can now use, because we use standrd urls?

File: frontend/packages/kubevirt-plugin/console-extensions.json (hunk 9 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/console-extensions.json:123]: let's do in a followup

File: frontend/packages/kubevirt-plugin/src/components/modals/menu-actions-modals/utils.ts (hunk 11 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/components/modals/menu-actions-modals/utils.ts:6]: let's do in a followup

File: frontend/public/components/app-contents.tsx [hunk 12]
```diff
@@ -704,7 +701,7 @@ const AppContents: React.FC<{}> = () => {
               path="/k8s/ns/:ns/pods/:podName/containers/:name"
               loader={() => import('./container').then((m) => m.ContainersDetailsPage)}
             />
-            <Route path="/k8s/ns/:ns/:plural/~new" exact component={CreateResource} />
```
Comment (yaacov) [inline frontend/public/components/app-contents.tsx:707]: don't you have a different PR for that ?

Comment (metalice) [reply frontend/public/components/app-contents.tsx:707]: https://github.com/openshift/console/pull/10488 
Im hoping when ^^ will merged, this file will automatically be removed as it will be the same as master, if not once it merged ill remove it

Comment (metalice) [reply frontend/public/components/app-contents.tsx]: didnt waited, removed it

File: frontend/packages/kubevirt-plugin/console-extensions.json (hunk 9 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/console-extensions.json]: /~new for vm resource is done in this pr

File: frontend/packages/kubevirt-plugin/console-extensions.json (hunk 4 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/console-extensions.json]: changed to  ... have templates for...

File: frontend/packages/kubevirt-plugin/console-extensions.json (hunk 3 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/console-extensions.json]: done

File: frontend/packages/kubevirt-plugin/src/constants/url-params.ts [hunk 13]
```diff
@@ -1,6 +1,26 @@
...
+
 export enum VMWizardURLParams {
   NAMESPACE = 'namespace',
   MODE = 'mode',
   INITIAL_DATA = 'initialData',
   VIEW = 'view',
 }
+
+const baseURLBuilder = (namespace: string | undefined) =>
+  `/k8s/ns/${namespace || 'default'}/${kubevirtReferenceForModel(VirtualMachineModel)}/~new`;
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/constants/url-params.ts:12]:
```suggestion
const baseURLBuilder = (namespace: string | undefined) =>
resourceListPathFromModel(getKubevirtAvailableModel(VirtualMachineModel), namespace)).concat('~new')
```

File: frontend/packages/kubevirt-plugin/src/components/VirtualizationWizardNavigator/VirtualizationWizardNavigator.tsx [hunk 14]
```diff
@@ -0,0 +1,20 @@
+import * as React from 'react';
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/components/VirtualizationWizardNavigator/VirtualizationWizardNavigator.tsx:1]: nice

File: frontend/packages/kubevirt-plugin/src/constants/vm/constants.ts [hunk 15]
```diff
@@ -72,7 +72,7 @@ export const VIRTIO_WIN_IMAGE = 'virtio-win-image';
 
 export const PENDING_RESTART_LABEL = '(pending restart)';
 export const getPVCUploadURL = (pvcNamespace: string): string =>
-  `/k8s/ns/${pvcNamespace}/persistentvolumeclaims/~new/upload-form`;
+  `/k8s/ns/${pvcNamespace}/persistentvolumeclaims/~new/form`;
```
Comment (glekner) [inline frontend/packages/kubevirt-plugin/src/constants/vm/constants.ts:75]: why?
isn't `form` being used by the regular pvc form?

**Review by glekner — COMMENTED** [review]
nice work @metalice

File: frontend/packages/kubevirt-plugin/src/constants/url-params.ts (hunk 13 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/constants/url-params.ts:12]: cool, will check!

File: frontend/packages/kubevirt-plugin/src/components/VirtualizationWizardNavigator/VirtualizationWizardNavigator.tsx (hunk 14 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/components/VirtualizationWizardNavigator/VirtualizationWizardNavigator.tsx:1]: :)

File: frontend/packages/kubevirt-plugin/src/constants/vm/constants.ts (hunk 15 above)
Comment (metalice) [reply frontend/packages/kubevirt-plugin/src/constants/vm/constants.ts:75]: im not sure I understand what you mean, but this change is about making them consistent with all ~new console requirements
